Implement GPU signed scalar div, add scalar div to the hl api

agnesLeroy opened the pull request:
<!-- Feel free to delete the template if the PR (bumping a version e.g.) does not fit the template -->
closes: _please link all relevant issues_

### PR content/description

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+            let l = chosen_multiplier.l;
+
+            // SRA(n, l − 1)
+            let mut tmp = self.unchecked_scalar_right_shift(numerator, l - 1, streams);
```
We should be using `unchecked_scalar_right_shift_async`.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -32,6 +36,21 @@ impl CudaServerKey {
+        Scalar: ScalarMultiplier + DecomposableInto<u8> + CastInto<u64>,
+    {
+        let num_blocks = lhs.as_ref().d_blocks.lwe_ciphertext_count().0;
+        let mut result = self.extend_radix_with_sign_msb(lhs, num_blocks, streams);
```
Not directly related to this PR but still in the scope,`extend_radix_with_sign_msb` is not following the async/sync standards. There should exist `extend_radix_with_sign_msb_assign_async` and `extend_radix_with_sign_msb_async` so we use the latter here. The same for `signed_scalar_mul_high`, we should have an async variant and use it in `unchecked_signed_scalar_div`.

agnesLeroy replied:
I made the changes in this PR so everything is consistent.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+        }
+    }
+
+    pub fn signed_scalar_div<Scalar>(
```
We should have an async variant for this.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+        self.unchecked_signed_scalar_div(numerator, divisor, streams)
+    }
+
+    pub fn unchecked_signed_scalar_div_rem<Scalar>(
```
We should have an async variant for this.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+        (quotient, remainder)
+    }
+
+    pub fn signed_scalar_div_rem<Scalar>(
```
We should have an async variant for this.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+        self.unchecked_signed_scalar_div_rem(numerator, divisor, streams)
+    }
+
+    pub fn unchecked_signed_scalar_rem<Scalar>(
```
We should have an async variant for this.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_div_mod.rs:
```diff
@@ -403,4 +422,262 @@ impl CudaServerKey {
+        remainder
+    }
+
+    pub fn signed_scalar_rem<Scalar>(
```
We should have an async variant for this.

pdroalves commented on tfhe/src/integer/gpu/server_key/radix/scalar_shift.rs:
```diff
@@ -565,4 +565,76 @@ impl CudaServerKey {
+    ///
+    /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
+    ///   not be dropped until stream is synchronised
+    pub unsafe fn unchecked_scalar_right_shift_logical_assign_async<Scalar, T>(
```
Why this method is being added in this PR?

agnesLeroy replied:
It was convenient to add it, it's called in `unchecked_scalar_right_shift_logical_assign_async`.

pdroalves reviewed (changes requested):
Hey Agnes!

I was not able to review the HL API. Regarding the integer part, I missed the `assign_async`/`async` logic we always use for those operators.

agnesLeroy commented:
@pdroalves I modified the PR, I changed also the cast functions to have sync/async variants. The PR is significantly bigger now but I think this improves things a lot, thank you very much for your review.

tmontaigu commented:
The hlapi part looks good

pdroalves commented on tfhe/src/core_crypto/gpu/entities/lwe_ciphertext_list.rs:
```diff
@@ -56,7 +69,6 @@ impl<T: UnsignedInteger> CudaLweCiphertextList<T> {
         unsafe {
```
We don't need this unsafe clause anymore.

pdroalves commented on tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs:
```diff
@@ -135,6 +135,7 @@ pub trait SignedReciprocable:
     type DoublePrecision: DecomposableInto<u8>
         + ScalarMultiplier
         + CastFrom<<Self::Unsigned as Reciprocable>::DoublePrecision>
+        + CastInto<u64>
```
Do we want to modify CPU's code?

agnesLeroy replied:
I have to add it to implement scalar_mul_high actually.

pdroalves reviewed (changes requested):
Seems good! Just added two minor comments. There is a forgotten unsafe clause.
